File: doc/examples/DocExamples/ProductGuide/IdSite.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Stormpath.SDK.Account;
using Stormpath.SDK.Application;
using Stormpath.SDK.Auth;
using Stormpath.SDK.Client;
using Stormpath.SDK.CustomData;
using Stormpath.SDK.Directory;
using Stormpath.SDK.Http;
using Stormpath.SDK.IdSite;
using Stormpath.SDK.Oauth;
using Stormpath.SDK.Organization;

// ReSharper disable PossibleNullReferenceException
#pragma warning disable 219

namespace Stormpath.SDK.DocExamples.ProductGuide
{
    public class IdSite
    {
        public async Task BuildIdSiteUrl()
        {
            IClient client = null;

            #region code/csharp/idsite/build_idsite_url.cs

            var app = await client.GetApplicationAsync("application_url");

            var idSiteUrl = app.NewIdSiteUrlBuilder()
                .SetCallbackUri("http://mysite.foo/idsiteCallback")
                .Build();

            #endregion
        }

        public async Task ConsumeIdSiteAssertion()
        {
            IApplication app = null;

            #region code/csharp/idsite/build_idsite_url.cs

            var requestDescriptor = HttpRequests.NewRequestDescriptor()
                .WithMethod("GET")
                .WithUri("incoming_uri")
                .Build();

            var idSiteListener = app.NewIdSiteAsyncCallbackHandler(requestDescriptor);

            var accountResult = await idSiteListener.GetAccountResultAsync();

            #endregion
        }

        public async Task ExchangeIdSiteAssertionForToken()
        {
            IApplication app = null;
            string incoming_uri = null;

            #region code/csharp/idsite/jwt_for_oauth_req.cs

            var jwt = incoming_uri.Split('=')[1];

            var idSiteTokenExchangeRequest = OauthRequests
                .NewIdSiteTokenAuthenticationRequest()
                .SetJwt(jwt)
                .Build();

            var grantResponse = await app.NewIdSiteTokenAuthenticator()
                .AuthenticateAsync(idSiteTokenExchangeRequest);

            // Token is stored in grantResponse.AccessTokenString
#endregion
        }

        public void BuildLogoutUrl()
        {
            IApplication app = null;

            #region code/csharp/idsite/logout_from_idsite_req.cs

            var logoutUrl = app.NewIdSiteUrlBuilder()
                .SetCallbackUri("http://mysite.foo/idsiteCallback")
                .ForLogout()
                .Build();

            #endregion
        }

        public void ConsumeLogout()
        {
            IAccountResult accountResult = null;

            #region code/csharp/idsite/logout_from_idsite_resp.cs
            
            // In your callback handler, after GetAccountResultAsync()

            if (accountResult.Status == IdSiteResultStatus.Logout)
            {
                // This was a logout! Proceed accordingly...
            }

            #endregion
        }

        public void BuildPasswordResetUrl()
        {
            IApplication app = null;
            string sptoken_from_url = null;

            #region code/csharp/idsite/idsite_reset_pwd.cs

            var logoutUrl = app.NewIdSiteUrlBuilder()
                .SetCallbackUri("http://mysite.foo/idsiteCallback")
                .SetPath("/#/reset")
                .SetSpToken(sptoken_from_url)
                .Build();

            #endregion
        }
    }
}

File: doc/examples/DocExamplesVB/ProductGuide/IdSite.vb
Imports Stormpath.SDK.Application
Imports Stormpath.SDK.Client
Imports Stormpath.SDK.Http
Imports Stormpath.SDK.IdSite
Imports Stormpath.SDK.Oauth

Namespace DocExamplesVB.ProductGuide

    Public Class IdSite

        Public Async Function BuildIdSiteUrl() As Task
            Dim client As IClient = Nothing

#Region "code/vbnet/idsite/build_idsite_url.vb"

            Dim app = Await client.GetApplicationAsync("application_url")

            Dim idSiteUrl = app.NewIdSiteUrlBuilder() _
                .SetCallbackUri("http://mysite.foo/idsiteCallback") _
                .Build()

#End Region
        End Function

        Public Async Function ConsumeIdSiteAssertion() As Task
            Dim app As IApplication = Nothing

#Region "code/vbnet/idsite/build_idsite_url.vb"

            Dim requestDescriptor = HttpRequests.NewRequestDescriptor() _
                .WithMethod("GET") _
                .WithUri("incoming_uri") _
                .Build()

            Dim idSiteListener = app.NewIdSiteAsyncCallbackHandler(requestDescriptor)

            Dim accountResult = Await idSiteListener.GetAccountResultAsync()

#End Region
        End Function

        Public Async Function ExchangeIdSiteAssertionForToken() As Task
            Dim app As IApplication = Nothing
            dim incoming_uri as String = nothing

#Region "code/vbnet/idsite/jwt_for_oauth_req.vb"

            Dim jwt = incoming_uri.Split("="c)(1)

            Dim idSiteTokenExchangeRequest = OauthRequests _
                .NewIdSiteTokenAuthenticationRequest() _
                .SetJwt(jwt) _
                .Build()

            Dim grantResponse = Await app.NewIdSiteTokenAuthenticator() _
                .AuthenticateAsync(idSiteTokenExchangeRequest)

            ' Token is stored in grantResponse.AccessTokenString
#End Region
        End Function

        Public Sub BuildLogoutUrl()
            Dim app As IApplication = Nothing

#Region "code/vbnet/idsite/logout_from_idsite_req.vb"

            Dim logoutUrl = app.NewIdSiteUrlBuilder() _
                .SetCallbackUri("http://mysite.foo/idsiteCallback") _
                .ForLogout() _
                .Build()

#End Region
        End Sub

        Public Sub ConsumeLogout()
            Dim accountResult As IAccountResult = Nothing

            #Region "code/vbnet/idsite/logout_from_idsite_resp.vb"

            ' In your callback handler, after GetAccountResultAsync()

            If accountResult.Status = IdSiteResultStatus.Logout Then
                ' This was a logout! Proceed accordingly...
            End If

            #End Region
        End Sub
        
Public Sub BuildPasswordResetUrl()
	Dim app As IApplication = Nothing
           Dim sptoken_from_url As String = nothing

	#Region "code/vbnet/idsite/idsite_reset_pwd.vb"

	Dim logoutUrl = app.NewIdSiteUrlBuilder() _
                .SetCallbackUri("http://mysite.foo/idsiteCallback") _
                .SetPath("/#/reset") _
                .SetSpToken(sptoken_from_url) _
                .Build()

	#End Region
End Sub

    End Class
End Namespace
